src/com/facebook/buck/swift/SwiftDescriptions.java

```diff
@@ -53,73 +40,47 @@
         swiftSrcsBuilder.add(source.getSourcePath());
       }
     }
-    ImmutableSortedSet<SourcePath> swiftSrcs = swiftSrcsBuilder.build();
-    if (swiftSrcs.isEmpty()) {
-      return Optional.absent();
-    }
-
-    AppleCxxPlatform appleCxxPlatform = ApplePlatforms.getAppleCxxPlatformForBuildTarget(
-        cxxPlatformFlavorDomain,
-        defaultCxxPlatform,
-        platformFlavorsToAppleCxxPlatforms,
-        swiftCompanionTarget,
-        Optional.<MultiarchFileInfo>absent());
-    Optional<Tool> swiftCompiler = appleCxxPlatform.getSwift();
-    if (!swiftCompiler.isPresent()) {
-      throw new HumanReadableException("Platform %s is missing swift compiler", appleCxxPlatform);
-    }
-
-    CxxPlatform cxxPlatform = cxxPlatformFlavorDomain.getValue(swiftCompanionTarget)
-        .or(defaultCxxPlatform);
+    return swiftSrcsBuilder.build();
+  }
 
-    Collection<CxxPreprocessorInput> cxxPreprocessorInputs =
-        CxxPreprocessables.getTransitiveCxxPreprocessorInput(cxxPlatform, parentParams.getDeps());
+  public static <A extends CxxLibraryDescription.Arg> void populateSwiftLibraryDescriptionArg(
+      final SourcePathResolver sourcePathResolver,
+      SwiftLibraryDescription.Arg output,
+      final A args,
+      BuildTarget buildTarget) {
+    output.srcs = args.srcs.transform(
+        new Function<ImmutableSortedSet<SourceWithFlags>, ImmutableSortedSet<SourcePath>>() {
+          @Override
+          public ImmutableSortedSet<SourcePath> apply(ImmutableSortedSet<SourceWithFlags> input) {
+            return filterSwiftSources(sourcePathResolver, args.srcs.get());
+          }
+        });
+    output.compilerFlags = args.compilerFlags;
+    output.frameworks = args.frameworks;
+    output.libraries = args.libraries;
+    output.deps = args.deps;
+    output.supportedPlatformsRegex = args.supportedPlatformsRegex;
+    output.moduleName = Optional.of(buildTarget.getShortName());
+    output.enableObjcInterop = Optional.of(true);
+  }
 
-    BuildRuleParams params = parentParams.copyWithBuildTarget(swiftCompanionTarget);
-    return Optional.<BuildRule>of(new SwiftLibrary(
-        swiftCompiler.get(),
-        params,
-        sourcePathResolver,
-        ImmutableList.<BuildRule>of(),
-        args.frameworks.get(),
-        args.libraries.get(),
-        platformFlavorsToAppleCxxPlatforms,
-        cxxPreprocessorInputs,
-        BuildTargets.getGenPath(
-            params.getProjectFilesystem(),
-            swiftCompanionTarget, "%s"),
-        swiftCompanionTarget.getShortName(),
-        swiftSrcs,
-        Optional.of(true),
-        Optional.<Pattern>absent()));
+  static boolean hasSwiftSource(ImmutableSortedSet<SourceWithFlags> srcs) {
+    return FluentIterable.from(srcs)
+        .firstMatch(new Predicate<SourceWithFlags>() {
+          @Override
+          public boolean apply(SourceWithFlags input) {
+            return MorePaths.getFileExtension(Paths.get(input.getSourcePath().toString()))
```

Comment:
You need to `sourcePathResolver.getAbsolutePath(source.getSourcePath())` here, like above in `filterSwiftSources`.

In fact since it's the same, extract the common logic.